waspc/src/Wasp/Analyzer/Parser/Lexer.x

```diff
@@ -97,6 +98,17 @@ alexGetByte (_, [], (currChar:remainingSource)) = case encodeChar currChar of
 alexInputPrevChar :: AlexInput -> Char
 alexInputPrevChar (prevChar, _, _) = prevChar
 
+-- | Tokenizes the given string. 
+alexScanTokens :: String -> Maybe [Parser Token]
```

Comment:
Usually stuf generated by Alex is prefixed with `alex`. So maybe we should name it `lexerScanTokens` instead? Or `lexerTokenize`? I wonder what is the difference, semantically, between this function and the `lexer` function below. `lexer` function below scans the next token and then gives control to parser. So, it scans only one token. Actually I am looking at its comments, it says exactly that, `Lexes a single token from the input`. This one, however, scans many tokens at once, without invoking any outside mechanisms (e.g. parser).

Actually, what `lexer` does is not only scan that single token, but also parse it -> since it calls `parseToken` function.

On the other hand, our new function scans all the tokens, and it doesn't parse them (although it operates in the Parser monad which is a bit confusing but ok).

So what about naming? Should we rename `lexer` to `lexAndParseSingleToken`?
Should we rename `alexScanTokens` to `lexAllTokens`?
Or maybe `lexerScanAndParseSingleToken` and `lexerTokenize`?
Something in that direction maybe.

Reply:
> Usually stuff generated by Alex is prefixed with alex. So maybe we should name it lexerScanTokens instead? 

Good point. I named it like this because that's what Alex did in [their documentation](https://www.haskell.org/alex/doc/html/wrappers.html), but I've broken the expected type of that function anyway (which is perhaps even a stronger argument for renaming it).

BTW, turns out their documentation _is_ up to date after all, but they also host the [old version](https://www.haskell.org/alex/doc/html-OLD/wrappers.html) which apparently has better SEO than the new one :)). 

Now, about the naming.

I think it makes sense to keep `lexer` as is - it's pretty standard and the name implies that the parser uses it (which is true). We do have to fix the documentation though, I'll make sure to do that.

I think my favorite options for `alexScanTokens` are `lexerTokenize` or just `tokenize`? I still hate the fact it returns `Maybe [Parser Token`]. It would be much nicer if I could get rid of the `Parser` monad before returning (making it return `Maybe [Token]`, but it's probably best to keep as much code as possible in normal `.hs` files (and avoid circular dependencies).

Reply:
I get what you mean with `Parser`, but probably it is worth investing time right now into trying to avoid it.
Naming -> ok, well the main thing I wanted to achieve was to make it clear what is the difference between `lexer` and `alexScanTokens`. In a sense, they are equals - they are both top level lexer functions. Which is why `lexerScanAndParseSingleToken` and `lexerScanAllTokens` made sense to me. Having the first one named just `lexer` somehow indicates it is a bit special, which it is, but it also looses the info on what it actually does. RIght now we do have that info in comments, but as the saying goes, if you can document something with names or types instead of comments, that is usually a good idea.
Ok, as long as renamed `alexScanTokens` to something else, I am good! I would at least try to keep it consistent with `lexer` somehow, as they really are two very similar functions, just serving different purposes / offering a bit different interface / behaviour.

Reply:
> I get what you mean with Parser, but probably it is **[not]** worth investing time right now into trying to avoid it.

Is this what you meant to say? :smile: 

The naming arguments make sense, I'll rename them both to something sensible.